include/aws/common/task_scheduler.h

```diff
@@ -0,0 +1,101 @@
+#ifndef AWS_TASK_SCHEDULER_H
+#define AWS_TASK_SCHEDULER_H
+
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/priority_queue.h>
+#include <stdint.h>
+
+typedef void *aws_task_arg;
+
+/**
+ * Pointer to the scheduled function
+ */
+typedef void(*aws_task_fn)(aws_task_arg arg);
+
+/**
+ * Monotonic clock function
+ */
+typedef int(*aws_task_scheduler_clock)(uint64_t *);
+
+struct aws_task {
+    aws_task_fn fn;
+    aws_task_arg arg;
+};
+
+struct aws_task_scheduler {
+    struct aws_allocator *alloc;
+    struct aws_priority_queue queue;
+    uint64_t owning_thread;
+    aws_task_scheduler_clock clock;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a task scheduler instance with a clock function and config. A priority queue will be allocated 
+     * in dynamic mode with initial size of 7
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
+     */
+    AWS_COMMON_API int aws_task_scheduler_init_default(struct aws_task_scheduler *scheduler, 
```

Comment:
We keep writing into preallocated structures everywhere. Are we really okay with throwing ABI forward-compatibility to the wind?

Reply:
I don't think it's necessarily being thrown to the wind, there's other mechanisms for establishing ABI compat than heap allocating everything.

Besides, it's not like the problem magically goes away with a _new call anyways, you'd still encounter different libs making the same call. You'd still have to do size checks everywhere. 

I'd rather nail these data structure APIs and not need to change them than start the habbit of forcing everything to dynamic memory and fragmented allocations all the way at the bottom of the dependency tree.